cdap-watchdog/src/main/java/co/cask/cdap/logging/write/FileMetaDataManager.java

```diff
@@ -138,26 +138,25 @@ public int cleanMetaData(final long tillTime, final DeleteCallback callback) thr
     return execute(new TransactionExecutor.Function<Table, Integer>() {
       @Override
       public Integer apply(Table table) throws Exception {
-        byte[] tillTimeBytes = Bytes.toBytes(tillTime);
-
         int deletedColumns = 0;
         Scanner scanner = table.scan(ROW_KEY_PREFIX, ROW_KEY_PREFIX_END);
         try {
           Row row;
           while ((row = scanner.next()) != null) {
             byte[] rowKey = row.getRow();
             String namespacedLogDir = LoggingContextHelper.getNamespacedBaseDir(logBaseDir, getLogPartition(rowKey));
-            byte[] maxCol = getMaxKey(row.getColumns());
 
             for (Map.Entry<byte[], byte[]> entry : row.getColumns().entrySet()) {
               byte[] colName = entry.getKey();
               if (LOG.isDebugEnabled()) {
                 LOG.debug("Got file {} with start time {}", Bytes.toString(entry.getValue()),
                           Bytes.toLong(colName));
               }
-              // Delete if colName is less than tillTime, but don't delete the last one
-              if (Bytes.compareTo(colName, tillTimeBytes) < 0 && Bytes.compareTo(colName, maxCol) != 0) {
-                callback.handle(locationFactory.create(new URI(Bytes.toString(entry.getValue()))), namespacedLogDir);
+
+              Location fileLocation = locationFactory.create(new URI(Bytes.toString(entry.getValue())));
+              // Delete if file last modified time is less than tillTime
+              if (fileLocation.lastModified() < tillTime) {
```

Comment:
Since the time in the meta table is the time of the first event, we saw a case where files were getting deleted earlier than expected when using the first event time for deletion. 

For now this is the best fix I could come up with that will not delete file sooner than is expected. I verified on a cluster that deletion based on the file modified time works.